controllers/module_nmc_reconciler.go

```diff
@@ -0,0 +1,150 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/filter"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/module"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/nmc"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//+kubebuilder:rbac:groups="core",resources=nodes,verbs=get;watch;patch
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs,verbs=get;list;watch;update;patch;create
+
+const (
+	ModuleNMCReconcilerName = "ModuleNMCReconciler"
+)
+
+type ModuleNMCReconciler struct {
+	kernelAPI   module.KernelMapper
+	filter      *filter.Filter
+	reconHelper moduleNMCReconcilerHelperAPI
+	nmcHandler  nmc.ModuleNMCHandler
+}
+
+func NewModuleNMCReconciler(client client.Client,
+	kernelAPI module.KernelMapper,
+	nmcHandler nmc.ModuleNMCHandler,
+	filter *filter.Filter) *ModuleNMCReconciler {
+	reconHelper := newModuleNMCReconcilerHelper(client)
+	return &ModuleNMCReconciler{
+		kernelAPI:   kernelAPI,
+		filter:      filter,
+		reconHelper: reconHelper,
+		nmcHandler:  nmcHandler,
+	}
+}
+
+func (mnr *ModuleNMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// get the reconciled module
+	logger := log.FromContext(ctx)
+
+	logger.Info("Starting Module-NMS reconcilation", "module name and namespace", req.NamespacedName)
+
+	mod, err := mnr.reconHelper.getRequestedModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("Module deleted")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get the requested %s KMMO CR: %w", req.NamespacedName, err)
+	}
+
+	// get all nodes
+	nodes, err := mnr.reconHelper.getNodesList(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get nodes list: %v", err)
+	}
+
+	errs := make([]error, 0, len(nodes))
+	for _, node := range nodes {
+		kernelVersion := strings.TrimSuffix(node.Status.NodeInfo.KernelVersion, "+")
+		mld, err := mnr.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
+		if err != nil && !errors.Is(err, module.ErrNoMatchingKernelMapping) {
+			logger.Info(utils.WarnString(fmt.Sprintf("internal errors while fetching kernel mapping for version %s: %v", kernelVersion, err)))
+			errs = append(errs, err)
+			continue
+		}
+
+		err = mnr.nmcHandler.HandleModuleNMC(ctx, mod, mld, node)
+		errs = append(errs, err)
```

Comment:
Aren't you appending `nil` errors here as well?

Reply:
yes, but errors.Join knows how to handle them and ignore

Reply:
Won't it be clearer to only append real errors instead of delegating it to `Join`?

Reply:
why to add another "if", if there is already code in the Join that takes care of it? This is the purpose of Join function: decide of there are errors in the array of errors , and create the appropriate error

Reply:
> why to add another "if", if there is already code in the Join that takes care of it?

Because it is not trivial for someone who didn't read the `Join` doc to understand those are ignored. Everywhere else in the code we are checking if an error is != `nil` before doing something with it except here.

If you think it is better this way, let's keep it, I just though it is a bit confusing.

Reply:
i am keeping it: if we have a Golang library function in the code, i would expect that developer to read about it and understand